ast/parser.go

```diff
@@ -713,6 +713,10 @@ func (p *Parser) parseRules() []*Rule {
 
 	// p[x] if ...  becomes a single-value rule p[x]
 	if hasIf && !usesContains && len(rule.Head.Ref()) == 2 {
+		if !rule.Head.Ref()[1].IsGround() && len(rule.Head.Args) == 0 {
+			rule.Head.Key = rule.Head.Ref()[1]
+		}
```

Comment:
Pulled from https://github.com/open-policy-agent/opa/pull/7004

Without this change, `topdown.TestTopDownPartialEval` will fail on rule head comparisons for the following test cases:

* `automatic shallow inlining: full extent: partial object`
* `automatic shallow inlining: full extent: partial object, general ref head`
* `automatic shallow inlining: full extent: iteration`
* `ref heads: unknown ref var, unknown rule value, shallow inlining`
* `ref heads: unknown ref var, unknown rule value, shallow inlining#01`

Modifications in this PR outside of the `topdown` package are related to this.